benchmark/blobs_benchmark.py

```diff
@@ -0,0 +1,258 @@
+    weighted_mmd = float(weighted_coresubset.compute_metric(mmd_metric))
+    weighted_ksd = float(weighted_coresubset.compute_metric(ksd_metric))
+
+    end_time = time.perf_counter()
+    elapsed_time = end_time - start_time
+
+    return {
+        "unweighted_mmd": unweighted_mmd,
+        "unweighted_ksd": unweighted_ksd,
+        "weighted_mmd": weighted_mmd,
+        "weighted_ksd": weighted_ksd,
+        "time": elapsed_time,
+    }
+
+
+def compute_metrics(
+    solvers: list[tuple[str, Any]],
+    dataset: Data,
+    mmd_metric: MMD,
+    ksd_metric: KSD,
+    weights_optimiser: MMDWeightsOptimiser,
+) -> dict[str, dict[str, float]]:
+    """
+    Compute the coresubsets and corresponding metrics for each solver in a given list.
+
+    :param solvers: A list of tuples containing solver names and their
+                    respective solver objects.
+    :param dataset: The dataset.
+    :param mmd_metric: The MMD metric object for computing Maximum Mean Discrepancy.
+    :param ksd_metric: The KSD metric object for computing Kernel Stein Discrepancy.
+    :param weights_optimiser: The optimizer object for weights for the coresubset.
+
+    :return: A dictionary where the keys are the solver names, and the values are
+             dictionaries of computed metrics (unweighted/weighted MMD and KSD, and
+             computation time).
+    """
+    return {
+        name: compute_solver_metrics(
+            solver, dataset, mmd_metric, ksd_metric, weights_optimiser
+        )
+        for name, solver in solvers
+    }
+
+
+def main() -> None:
+    """
+    Benchmark different algorithms against on a synthetic dataset.
+
+    Compare the performance of different coreset algorithms using a synthetic dataset,
+    generated using :func:`sklearn.datasets.make_blobs`. We set up various solvers,
+    generate coresets of multiple sizes, and compute performance metrics (MMD and KSD)
+    for each solver at each coreset size. Results are saved to a JSON file.
+    """
+    # Generate data
+    x, *_ = make_blobs(n_samples=1000, n_features=2, centers=10, random_state=45)
+    dataset = Data(jnp.array(x))
+
+    # Set up kernel
+    sq_exp_kernel = setup_kernel(x)
+
+    # Set up Stein Kernel
+    stein_kernel = setup_stein_kernel(sq_exp_kernel, dataset)
+
+    # Set up metrics
+    mmd_metric = MMD(kernel=sq_exp_kernel)
+    ksd_metric = KSD(kernel=sq_exp_kernel)
+
+    # Set up weights optimizer
+    weights_optimiser = MMDWeightsOptimiser(kernel=sq_exp_kernel)
+
+    # Define coreset sizes
+    coreset_sizes = [10, 50, 100, 200]
+
+    all_results = {}
+
+    for size in coreset_sizes:
+        solvers = setup_solvers(size, sq_exp_kernel, stein_kernel)
+
+        # Compute metrics
+        results = compute_metrics(
+            solvers, dataset, mmd_metric, ksd_metric, weights_optimiser
+        )
+        all_results[size] = results
+
+    # Save results to JSON file
+    with open("blobs_benchmark_results.json", "w", encoding="utf-8") as f:
```

Comment:
Add more control over where this results file is saved. Suggested change:

    base_dir = os.path.dirname(os.path.abspath(__file__))  # Gets the directory of current script
    # Save results to JSON file
    with open(os.path.join(base_dir,"blobs_benchmark_results.json"), "w", encoding="utf-8") as f: